docs/architecture/adr-006-upgrade-client-implementation.md

```diff
@@ -0,0 +1,433 @@
+          `TrustingPeriod`).
+   2. Vote on and pass the `UpgradeProposal`
+   3. Commit `UpgradedClient` by the upgrade module under the following key:
+
+        ```md
+        upgrade/UpgradedIBCState/{upgradeHeight}/upgradedClient
+        ```
+
+   4. Commit an initial consensus state by upgrade module on the block right
+      before the upgrade height for the next chain under the following key:
+
+        ```md
+        upgrade/UpgradedIBCState/{upgradeHeight}/upgradedConsState
+        ```
+
+2. Submit an upgrade client message by relayers to the counterparty chain
+   1. Wait for the upgrading chain to reach the upgrade height and halt
+   2. Query a full node for the proofs of `UpgradedClient` and
+      `UpgradedConsensusState` at the last height of the old chain
+   3. Update the counterparty client to the last height of the old chain using
+      the `UpdateClient` msg
+   4. Submit an `UpgradeClient` msg to the counterparty chain with the
+      `UpgradedClient`, `UpgradedConsensusState` and their respective proofs
+
+3. Process the upgrade message on the counterparty chain</br> IBC handler upon
+receiving a `MsgUpgradeClient` message performs basic validations (BV),
+client-specific validations (SV) and lastly execution (E) steps as follows:
+   1. (BV) Check that the current client is not frozen
+   2. (BV) Check if the latest consensus state is within the trust period
+   3. (BV) Check if the message containing proofs decoded successfully
+   4. (SV) Verify that the upgradedClient be of a Tendermint `ClientState` type
+   5. (SV) Verify that the upgradedConsensusState be of a Tendermint
+      `ConsensusState` type
+   6. (SV) Check the latest height of the current client state does not have the
+      same revision number or has a greater height than the committed client
+   7. (SV) Match any Tendermint chain specified parameter in upgraded client
+      such as ChainID, UnbondingPeriod, and ProofSpecs with the committed client
+   8. (SV) Verify that the upgrading chain did indeed commit to the upgraded
+      client state at the upgrade height by provided proof
+      Note: client-customizable fields must be zeroed out for this check
+   9. (SV) Verify that the upgrading chain did indeed commit to the upgraded
+      consensus state at the upgrade height by provided proof
+   10. (E) Upgrade client to the new client by retaining old client-customizable
+      parameters (sent by relayers) such `TrustingPeriod`, `TrustLevel`,
+      `MaxClockDrift` and adopt the new chain-specified fields such as
+      `UnbondingPeriod`, `ChainId`, `UpgradePath`, etc.
+   11. (E) Upgrade consensus state with a stand-in sentinel value for root</b>
+      Note: The upgraded consensus state serve purely as a basis of trust for
+      future `UpdateClientMsgs`, and therefore does not require a root for proof
+      verification. Also, we do not set processed time for this consensus state
```

Comment:
Can you expand on this? What's a "stand-in sentinel value"? How does it serve as a basis of trust for future `UpdateClient` msgs? That is, how is the next `UpdateClient` message going to use the value?

Reply:
After thinking about it some more, I get it, and it's actually quite simple 😅. I would call it a "dummy value", as it will never be used for packet verification or anything else. Basically the only fields that matter is the validator hash ones, and the one needed for checking the trusting period. I guess it just makes intuitive sense for me now why putting a dummy value as the root will only affect Merkle proofs, and not the ability to accept a new valid header with a `MsgUpdateClient`. All good!